src/tactic/rewrite_search/search.lean

```diff
@@ -0,0 +1,212 @@
+new vertices and edges.
+-/
+meta def mk_graph (conf : config) (rules : list (expr × bool)) (eq : expr) : tactic graph :=
+do (lhs, rhs) ← split_equation eq,
+  lhs_pp ← to_string <$> tactic.pp lhs,
+  rhs_pp ← to_string <$> tactic.pp rhs,
+  let lhs_vertex : vertex := ⟨0, lhs, lhs_pp, side.L, none⟩,
+  let rhs_vertex : vertex := ⟨1, rhs, rhs_pp, side.R, none⟩,
+  return ⟨conf, rules, [lhs_vertex, rhs_vertex].to_buffer,
+          native.rb_map.of_list [(lhs_pp, 0), (rhs_pp, 1)], none⟩
+
+variables (g : graph)
+
+namespace graph
+
+/--
+A convenience wrapper around `g.vertices` that fails if the index is out of bounds.
+-/
+meta def get_vertex (i : ℕ) : tactic vertex :=
+if h : i < g.vertices.size then return $ g.vertices.read (fin.mk i h)
+else fail "invalid vertex access"
+
+/--
+Find a list of edges that connect the given edge to the root of its tree.
+The edges are returned in leaf-to-root order, while they are in root-to-leaf direction,
+so if you want them in the logical order you must reverse the returned list.
+-/
+private meta def walk_up_parents : option edge → tactic (list edge)
+| none     := return []
+| (some e) := do
+  v ← g.get_vertex e.fr,
+  edges ← walk_up_parents v.parent,
+  return (e :: edges)
+
+/--
+Returns two lists that represent a solution. The first list is a path from LHS to some
+interior vertex, the second is a path from the RHS to that interior vertex.
+-/
+private meta def solution_paths : tactic (list edge × list edge) :=
+do e ← g.solving_edge,
+  v ← g.get_vertex e.to,
+  path1 ← walk_up_parents g e,
+  path2 ← walk_up_parents g v.parent,
+  match v.side with
+  | side.L := return (path2.reverse, path1.reverse)
+  | side.R := return (path1.reverse, path2.reverse)
+  end
+
+/--
+Add the new vertex and edge to the graph, that can be proved in one step starting
+at a given vertex, with a given rewrite expression.
+For efficiency, it's important that this is the only way the graph is mutated,
+and it only appends to the end of the `vertices` buffer.
+-/
+private meta def add_rewrite (v : vertex) (rw : rewrite) : tactic graph :=
+do pp ← to_string <$> tactic.pp rw.exp,
+  let maybe_id := g.vmap.find pp,
+  match maybe_id with
+  | (some id) := do
+    existing_vertex ← g.get_vertex id,
+    if v.side = existing_vertex.side then return g
+    else return { g with solving_edge := some ⟨v.id, existing_vertex.id, rw.proof, rw.how⟩ }
+  | none := do
+    let new_vertex_id := g.vertices.size,
+    let new_edge : edge := ⟨v.id, new_vertex_id, rw.proof, rw.how⟩,
+    let new_vertex : vertex := ⟨new_vertex_id, rw.exp, pp, v.side, (some new_edge)⟩,
+    trace_if_enabled `rewrite_search format!"new edge: {v.pp} → {new_vertex.pp}",
+    return { g with vertices := g.vertices.push_back new_vertex,
+                    vmap := g.vmap.insert pp new_vertex_id }
+end
+
+/--
+Add all single-step rewrites starting at a particular vertex to the graph.
+-/
+private meta def expand_vertex (v : vertex) : tactic graph :=
```

Comment:
For the most part this `search` and `discovery` are nicely modularized. This is the only declaration that breaks if `search` doesn't import `discovery`. Is it refactorable? Partly just leaving this as a note for myself as I continue to review.

Reply:
Okay, after reading more, it makes more sense that `discovery` is a module used by the search.